src/Razor/src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupStartTagSyntax.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+
+namespace Microsoft.AspNetCore.Razor.Language.Syntax
+{
+    internal partial class MarkupStartTagSyntax
+    {
+        public bool IsMarkupTransition
+        {
+            get
+            {
+                return ((InternalSyntax.MarkupStartTagSyntax)Green).IsMarkupTransition;
+            }
+        }
+
+        public SyntaxList<RazorSyntaxNode> Children => GetLegacyChildren();
+
+        public string GetTagName()
+        {
+            return Name.IsMissing ? string.Empty : Bang?.Content + Name.Content;
+        }
+
+        public bool IsSelfClosing()
+        {
+            return ForwardSlash != null &&
+                !ForwardSlash.IsMissing &&
+                !CloseAngle.IsMissing;
+        }
+
+        public bool IsVoidElement()
+        {
+            return ParserHelpers.VoidElements.Contains(GetTagName());
```

Comment:
This check doesn't fully work because of `bang`. Aka, for a `<!input... />` it wouldn't return true.